editor/plugins/curve_editor_plugin.cpp

```diff
@@ -311,236 +389,273 @@ void CurveEditor::on_preset_item_selected(int preset_id) {
+	undo_redo->add_undo_method(this, "set_selected_index", selected_index);
 	undo_redo->commit_action();
 }
 
-void CurveEditor::_curve_changed() {
+void CurveEdit::_curve_changed() {
 	queue_redraw();
-	// Point count can change in case of undo
-	if (_selected_point >= _curve_ref->get_point_count()) {
-		set_selected_point(-1);
+	// Point count can change in case of undo.
+	if (selected_index >= curve->get_point_count()) {
+		set_selected_index(-1);
 	}
 }
 
-void CurveEditor::on_context_menu_item_selected(int action_id) {
-	switch (action_id) {
-		case CONTEXT_ADD_POINT:
-			add_point(_context_click_pos);
-			break;
-
-		case CONTEXT_REMOVE_POINT:
-			remove_point(_selected_point);
-			break;
-
-		case CONTEXT_LINEAR:
-			toggle_linear();
-			break;
+int CurveEdit::get_point_at(Vector2 p_pos) const {
+	if (curve.is_null()) {
+		return -1;
+	}
 
-		case CONTEXT_LEFT_LINEAR:
-			toggle_linear(TANGENT_LEFT);
-			break;
+	const Rect2 hover_rect = Rect2(p_pos, Vector2(0, 0)).grow(hover_radius);
+	int closest_idx = -1;
+	float closest_dist_squared = hover_radius * hover_radius * 2;
 
-		case CONTEXT_RIGHT_LINEAR:
-			toggle_linear(TANGENT_RIGHT);
-			break;
+	for (int i = 0; i < curve->get_point_count(); ++i) {
+		Vector2 p = get_view_pos(curve->get_point_position(i));
+		if (hover_rect.has_point(p) && p.distance_squared_to(p_pos) < closest_dist_squared) {
+			closest_dist_squared = p.distance_squared_to(p_pos);
+			closest_idx = i;
+		}
 	}
-}
 
-void CurveEditor::open_context_menu(Vector2 pos) {
-	_context_menu->set_position(pos);
+	return closest_idx;
+}
 
-	_context_menu->clear();
+CurveEdit::TangentIndex CurveEdit::get_tangent_at(Vector2 p_pos) const {
+	if (curve.is_null() || selected_index < 0) {
+		return TANGENT_NONE;
+	}
 
-	if (_curve_ref.is_valid()) {
-		_context_menu->add_item(TTR("Add Point"), CONTEXT_ADD_POINT);
+	const Rect2 hover_rect = Rect2(p_pos, Vector2(0, 0)).grow(tangent_hover_radius);
 
-		if (_selected_point >= 0) {
-			_context_menu->add_item(TTR("Remove Point"), CONTEXT_REMOVE_POINT);
+	if (selected_index != 0) {
+		Vector2 control_pos = get_tangent_view_pos(selected_index, TANGENT_LEFT);
+		if (hover_rect.has_point(control_pos)) {
+			return TANGENT_LEFT;
+		}
+	}
 
-			if (_selected_tangent != TANGENT_NONE) {
-				_context_menu->add_separator();
+	if (selected_index != curve->get_point_count() - 1) {
+		Vector2 control_pos = get_tangent_view_pos(selected_index, TANGENT_RIGHT);
+		if (hover_rect.has_point(control_pos)) {
+			return TANGENT_RIGHT;
+		}
+	}
 
-				_context_menu->add_check_item(TTR("Linear"), CONTEXT_LINEAR);
+	return TANGENT_NONE;
+}
 
-				bool is_linear = _selected_tangent == TANGENT_LEFT
-						? _curve_ref->get_point_left_mode(_selected_point) == Curve::TANGENT_LINEAR
-						: _curve_ref->get_point_right_mode(_selected_point) == Curve::TANGENT_LINEAR;
+float CurveEdit::get_offset_without_collision(int p_current_index, float p_offset, bool p_prioritize_right) {
```

Comment:
> This also seems to be a loop that isn't necessarily guaranteed to finish? Or at the very least, a loop with an unknown, potentially large number of iterations, which feels risky.

> I believe what this is trying to do is pushing the location of a new point (under the mouse) away from existing points such that they no longer collide. Is that right?

You're correct about both.

I wouldn't remove it though, these bad scenarios won't occur in practice (I've strained it with thousands of points to make sure it doesn't hurt performance) and it helps to allow points get near each other's x-value with snapping or at the boundaries. For example, you can compare what happens with points at x=0 before and after this PR.

Reply:
Ok, so I ran a test, and even with a non-trivial amount of points all right next to each other (like around 15), the most number of iterations I found this took was ~68, which isn't really bad for an intentionally annoying made-up scenario. It doesn't seem, from my little testing, to go bonkers on the number of iterations! :)

If you add a lot of points to x=1 though, I did make it go up to ~630 iterations, and it goes up for every other point at x=1. People certainly won't intentional make this happen (what's the point of lots of points at x=1 anyway), but it does highlight the possibility for this to increase relatively easily under some (silly) conditions.

So I'm not too worried, but I'm also not super excited to see this in production code. Maybe an alternative algorithm is to: find the closest point to where the mouse is, and if there's a conflict, push the point directly away from the closest conflicting point by the length of the rect size that determines conflict. Repeat 2-3 times in case there are further conflicts, but then just pick that place. Then do some sanity checks (e.g. 0 <= x <= 1) and just accept that possibly imperfect spot?

I couldn't really find much difference between the current PR and 4.0.2 stable for points at x=0, but maybe I don't know what I'm looking for!